In [1]:
import sys
sys.path.append('../')
import latexStrings as ls
import numpy as np
import matplotlib.pyplot as plt
from math import *
from IPython.display import Latex
import odesolver

# Ejercicio 5

Tenemos el PVI del ejercicio 1:

\begin{equation} \label{eq:1}
y'= y + 2e^{-t} \qquad y(0) = 1 
\end{equation}

Dado que el lado derecho de la EDO, $y + 2e^{-t}$ es de clase $C^1$ para todo $(t,y)\in \mathbb{R}\times\mathbb{R}$, y por lo tanto Lipschitz Continua en dicha region, sabemos que existe una solucion unica a este problema, que en general es dada por:



 


In [2]:
Latex('$y(t) = (y_0e^{-t_0} + e^{-2t_0}-e^{-2t})e^t$')

<IPython.core.display.Latex object>

In [3]:
Latex('$$\Rightarrow y(1)=(1e^{-0}+e^{-2(0)}-e^{-2(1)})e^1=5.068684215746647$$')

<IPython.core.display.Latex object>

Ahora utilizemos los métodos:
    Euler (explícito),
    Trapecio (explícito), Punto Medio (explícito) 
    y RK4 con h ∈ {1/50,1/100,1/200} para calcular el error global en t=1

In [4]:
EG=[[0,0,0],[0,0,0],[0,0,0],[0,0,0]]
f = lambda t, y : y+2*np.exp(-t)
exact = lambda t, t0, y0 : np.exp(t)*(y0*np.exp(-t0)+np.exp(-2*t0)-np.exp(-2*t))
I = (0, 1)
y0 = 1
y1 = exact(1,0,1)
TEE, WEE = odesolver.solve(f,y0,I,50,method="Explicit Euler")
EG[0][0] = abs(WEE[0,50]-y1)
TEE1, WEE1 = odesolver.solve(f,y0,I,100,method="Explicit Euler")
EG[0][1] = abs(WEE1[0,100]-y1)
TEE2, WEE2 = odesolver.solve(f,y0,I,200,method="Explicit Euler")
EG[0][2] = abs(WEE2[0,200]-y1)


In [5]:
TTE, WTE = odesolver.solve(f,y0,I,50,method="Explicit Trapezoid")
EG[1][0]= abs(WTE[0,50]-y1)
TTE1, WTE1 = odesolver.solve(f,y0,I,100,method="Explicit Trapezoid")
EG[1][1] = abs(WTE1[0,100]-y1)
TTE2, WTE2 = odesolver.solve(f,y0,I,200,method="Explicit Trapezoid")
EG[1][2] = abs(WTE2[0,200]-y1)


In [6]:
TPE, WPE = odesolver.solve(f,y0,I,50,method="Explicit Midpoint")
EG[2][0]= abs(WPE[0,50]-y1)
TPE1, WPE1 = odesolver.solve(f,y0,I,100,method="Explicit Midpoint")
EG[2][1] = abs(WPE1[0,100]-y1)
TPE2, WPE2 = odesolver.solve(f,y0,I,200,method="Explicit Midpoint")
EG[2][2] = abs(WPE2[0,200]-y1)


In [7]:
TRK, WRK = odesolver.solve(f,y0,I,50,method="RK4")
EG[3][0]= abs(WRK[0,50]-y1)
TRK1, WRK1 = odesolver.solve(f,y0,I,100,method="RK4")
EG[3][1] = abs(WRK1[0,100]-y1)
TRK2, WRK2 = odesolver.solve(f,y0,I,200,method="RK4")
EG[3][2]= abs(WRK2[0,200]-y1)


In [8]:
EG

[[0.04178851634198466, 0.021098542854763203, 0.01060108612541466],
 [0.00020152521281335822, 5.09060053159871e-05, 1.2792587365417774e-05],
 [0.0003178642613095306, 8.013876027668232e-05, 2.0119208638647024e-05],
 [4.13889189587735e-09, 2.6199131752946414e-10, 1.6481038755955524e-11]]

En la siguiente tabla podemos observar el error global en cada uno de los casos y comparalo con el orden de los métodos:

$
\begin{table}[htb]
\centering
\begin{tabular}{|l|l|l|l|l|}
\hline
\cline{2-5}
& h & $\frac{1}{50}$ & $\frac{1}{100}$ &  $\frac{1}{200}$\\
\hline 
\multirow{2}{4cm}{Euler Explícito } & Error Global  & 0.04178851634198466 & 0.021098542854763203 & 0.01060108612541466\\ \cline{2-5}
& Orden 1 & O(0.02) & O(0.01)  & O(0.005) \\ \cline{1-5}
\multirow{2}{4cm}{Trapecio Explícito} & Error Global  & 2.0152521281335822e-04 &  5.09060053159871e-05 & 1.2792587365417774e-05 \\ \cline{2-5}
& Orden 2 & O(4e-04) & O(e-04) & O(2.5e-05)\\ \cline{1-5}
\multirow{2}{4cm}{Punto Medio Explícito} & Error Global  & 3.178642613095306e-04 & 8.013876027668232e-05 & 2.0119208638647024e-05\\ \cline{2-5}
& Orden 2 &  O(4e-04) & O(e-04) & O(2.5e-05)\\ \cline{1-5}
\multirow{2}{4cm}{RK4 } & Error Global  & 4.13889189587735e-09 & 2.6199131752946414e-10 & 1.6481038755955524e-11\\ \cline{2-5}
& Orden 4 & O(1.6e-07) & O(e-08) & O(6.25e-10)\\ \cline{1-5}
\end{tabular}
\caption{Error Global}
\label{tabla:final}
\end{table}
$